packages/doenetml-worker-rust/doenetml-core/src/state_var_calculations.rs

```diff
@@ -0,0 +1,762 @@
+                // and value from val_idx, we initially skip val_idx values
+                // and then set initial_val_idx to zero to not skip values on further instructions.
+                let mut initial_val_idx = stale_state.val_idx;
+
+                let dependencies_for_state_var =
+                    &dependencies[component_state.component_idx][component_state.state_var_idx];
+
+                for (instruction_idx, deps) in dependencies_for_state_var
+                    .iter()
+                    .enumerate()
+                    .skip(stale_state.instruction_idx)
+                {
+                    for (val_idx, dep) in deps.iter().enumerate().skip(initial_val_idx) {
+                        if let DependencySource::StateVar {
+                            component_idx,
+                            state_var_idx,
+                        } = &dep.source
+                        {
+                            let new_component_state = ComponentStateDescription {
+                                component_idx: *component_idx,
+                                state_var_idx: *state_var_idx,
+                            };
+
+                            let new_current_freshness = dep.value.get_freshness();
+
+                            match new_current_freshness {
+                                // No need to recurse if the state var of the dependency is already fresh
+                                // so don't do anything to just continue the inner loop
+                                Freshness::Fresh => (),
+                                Freshness::Stale => {
+                                    // Recurse by putting two items on the stack
+                                    // and then continuing at start of stack loop
+                                    freshen_stack.push(StateVarCalculationState::Stale(
+                                        StaleCalculationState {
+                                            component_state,
+                                            instruction_idx,
+                                            val_idx,
+                                        },
+                                    ));
+                                    freshen_stack.push(StateVarCalculationState::Stale(
+                                        StaleCalculationState {
+                                            component_state: new_component_state,
+                                            instruction_idx: 0,
+                                            val_idx: 0,
+                                        },
+                                    ));
+
+                                    // break out of the inner loop and continue to outer loop,
+                                    // which will result in first popping off the last instruction
+                                    // that we added to the stack
+                                    continue 'stack_loop;
```

Comment:
I looks like you're doing a depth-first search for stale deps. If you did a breadth-first search could you avoid needing to break out of this double loop?

Reply:
I need to do a depth-first search because I can calculate/freshen a variable only after all its dependencies (the subgraph below it) have been freshened.

Reply:
I don't understand...If you add each depth-1 state far to the list, next it will pick the last depth-1 state var and all the depth-2 state var deps to the list, etc. All deps should be handled in the correct order.

Reply:
Have a much simpler version with depth-first now.